### ISSUE TITLE
ONNX model produces different results in tensorRT 7 compared to pytorch and tensorRT 6

### ISSUE THREAD

**Issue by taxuezcy** [NVIDIA/TensorRT#1087]
## Environment

**TensorRT Version**: 7
**NVIDIA GPU**: 
**NVIDIA Driver Version**: 
**CUDA Version**: 10.2
**CUDNN Version**: 7.6.5
**Operating System**: 
**Python Version (if applicable)**: 3.6
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable)**: 1.6
**Baremetal or Container (if so, version)**: 

the same onnx model ,the result is different when use between trt6 and trt7, and the result in trt7 is very bad, it is a big bug,look forward for your reply

**Comment by taxuezcy** [comment]
@ttyio

**Comment by ttyio** [comment]
Hello @taxuezcy , sorry for the mismatch.
Could you check your ONNX model to see if there is `upsample` or `resize` node? thanks!

**Comment by taxuezcy** [comment]
> Hello @taxuezcy , sorry for the mismatch.
> Could you check your ONNX model to see if there is `upsample` or `resize` node? thanks!

hello, my onnx model do not have upsample or resize op ,but have ConvTranspose op , i don't know whether it is the problem

**Comment by taxuezcy** [comment]
@ttyio

**Comment by ttyio** [comment]
Hello @taxuezcy , There is no known issue for ConvTranspose.
Which precision are you using? could you share the onnx file and repro steps? 
Also there is a tool https://github.com/NVIDIA/TensorRT/tree/master/tools/Polygraphy you can try on your side, it support comparison between TRT and other DL framework. the command line looks like this:

         polygraphy run path_to_your.onnx --trt --onnxrt

**Comment by taxuezcy** [comment]
hi, polygraphy run result is 
[W] 'colored' module is not installed, will not use colors when logging. To enable colors, please install the 'colored' module: python3 -m pip install colored
[I] Runner: trt-runner-N0-02/26/21-09:16:03          | Activating and starting inference
[I] Building engine with configuration: max_workspace_size=16777216 bytes (16.00 MB) | fp16=False, int8=False, strict_types=False | 1 profiles
[TensorRT] WARNING: TensorRT was linked against cuBLAS 10.2.2 but loaded cuBLAS 10.1.0
[TensorRT] WARNING: TensorRT was linked against cuBLAS 10.2.2 but loaded cuBLAS 10.1.0
[TensorRT] WARNING: TensorRT was linked against cuBLAS 10.2.2 but loaded cuBLAS 10.1.0
[TensorRT] WARNING: Current optimization profile is: 0. Please ensure there are no enqueued operations pending in this context prior to switching profiles
[I] Runner: trt-runner-N0-02/26/21-09:16:03          | Completed 1 iterations.
[I] Runner: onnxrt-runner-N0-02/26/21-09:16:03       | Activating and starting inference
[I] Runner: onnxrt-runner-N0-02/26/21-09:16:03       | Completed 1 iterations.
[I] Accuracy Comparison | trt-runner-N0-02/26/21-09:16:03 vs. onnxrt-runner-N0-02/26/21-09:16:03
[I]     Comparing Output: 'output' (dtype=float32, shape=(1, 1, 128, 128)) with 'output' (dtype=float32, shape=(1, 1, 128, 128))
[I]         Required tolerances: [atol=1.7366] OR [rtol=1e-05, atol=1.7365] OR [rtol=0.44963, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-09:16:03          | Stats: mean=-5.3844, min=-7.9094 at (0, 0, 115, 3), max=-2.4424 at (0, 0, 126, 126)
            Runner: onnxrt-runner-N0-02/26/21-09:16:03       | Stats: mean=-5.3046, min=-8.0653 at (0, 0, 49, 9), max=-2.7 at (0, 0, 77, 0)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[I]     Comparing Output: '535' (dtype=float32, shape=(1, 2, 128, 128)) with '535' (dtype=float32, shape=(1, 2, 128, 128))
[I]         Required tolerances: [atol=0.23577] OR [rtol=1e-05, atol=0.23577] OR [rtol=5.8152, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-09:16:03          | Stats: mean=0.43141, min=-0.046996 at (0, 0, 50, 2), max=0.9565 at (0, 0, 35, 126)
            Runner: onnxrt-runner-N0-02/26/21-09:16:03       | Stats: mean=0.41433, min=-0.018458 at (0, 0, 50, 2), max=0.89764 at (0, 1, 121, 96)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[I]     Comparing Output: '536' (dtype=float32, shape=(1, 2, 128, 128)) with '536' (dtype=float32, shape=(1, 2, 128, 128))
[I]         Required tolerances: [atol=4.1119] OR [rtol=1e-05, atol=4.1117] OR [rtol=56.342, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-09:16:03          | Stats: mean=2.289, min=0.41181 at (0, 1, 127, 114), max=20.953 at (0, 1, 66, 1)
            Runner: onnxrt-runner-N0-02/26/21-09:16:03       | Stats: mean=2.2614, min=-0.12555 at (0, 0, 94, 0), max=23.085 at (0, 1, 65, 1)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[E]     FAILED | Mismatched outputs: ['output', '535', '536']
@ttyio

**Comment by ttyio** [comment]
Hello @taxuezcy , adding in polygraphy

      --trt-outputs mark all --onnx-outputs mark all 

can run the model with a layerwise comparison. Hopefully we can find out where the accuracy issue starts.

**Comment by taxuezcy** [comment]
hi, the mark all result is 

[I]     Comparing Output: '531' (dtype=float32, shape=(1, 24, 128, 128)) with '531' (dtype=float32, shape=(1, 24, 128, 128))
[I]         Required tolerances: [atol=5.579e-05] OR [rtol=1e-05, atol=3.1992e-05] OR [rtol=0.0019966, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-09:56:24          | Stats: mean=-0.16801, min=-23.276 at (0, 0, 32, 0), max=19.832 at (0, 16, 65, 1)
            Runner: onnxrt-runner-N0-02/26/21-09:56:24       | Stats: mean=-0.16801, min=-23.276 at (0, 0, 32, 0), max=19.832 at (0, 16, 65, 1)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[I]     Comparing Output: '532' (dtype=float32, shape=(1, 24, 128, 128)) with '532' (dtype=float32, shape=(1, 24, 128, 128))
[I]         Required tolerances: [atol=1.6212e-05] OR [rtol=1e-05, atol=1.2886e-05] OR [rtol=0.00049351, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-09:56:24          | Stats: mean=0.52479, min=-7.239 at (0, 0, 32, 0), max=5.6855 at (0, 1, 66, 0)
            Runner: onnxrt-runner-N0-02/26/21-09:56:24       | Stats: mean=0.52479, min=-7.239 at (0, 0, 32, 0), max=5.6855 at (0, 1, 66, 0)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[I]     Comparing Output: '533' (dtype=float32, shape=(1, 24, 128, 128)) with '533' (dtype=float32, shape=(1, 24, 128, 128))
[I]         Required tolerances: [atol=1.6212e-05] OR [rtol=1e-05, atol=1.2886e-05] OR [rtol=0.00049351, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-09:56:24          | Stats: mean=0.6944, min=0 at (0, 0, 0, 0), max=5.6855 at (0, 1, 66, 0)
            Runner: onnxrt-runner-N0-02/26/21-09:56:24       | Stats: mean=0.6944, min=0 at (0, 0, 0, 0), max=5.6855 at (0, 1, 66, 0)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[I]     Comparing Output: '534' (dtype=float32, shape=(1, 1, 128, 128)) with '534' (dtype=float32, shape=(1, 1, 128, 128))
[S]         PASSED | Difference is within tolerance (rtol=1e-05, atol=1e-05)
[I]     Comparing Output: '535' (dtype=float32, shape=(1, 2, 128, 128)) with '535' (dtype=float32, shape=(1, 2, 128, 128))
[S]         PASSED | Difference is within tolerance (rtol=1e-05, atol=1e-05)
[I]     Comparing Output: '536' (dtype=float32, shape=(1, 2, 128, 128)) with '536' (dtype=float32, shape=(1, 2, 128, 128))
[I]         Required tolerances: [atol=5.8651e-05] OR [rtol=1e-05, atol=1.9761e-05] OR [rtol=1.251e-05, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-09:56:24          | Stats: mean=2.2614, min=-0.12555 at (0, 0, 94, 0), max=23.085 at (0, 1, 65, 1)
            Runner: onnxrt-runner-N0-02/26/21-09:56:24       | Stats: mean=2.2614, min=-0.12555 at (0, 0, 94, 0), max=23.085 at (0, 1, 65, 1)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[E]     FAILED | Mismatched outputs: ['505', '507', '511', '513', '531', '532', '533', '536']

**Comment by ttyio** [comment]
@taxuezcy , after enable layerwise output, we can see the tolerance for `536` is fixed from  

      [rtol=1e-05, atol=4.1117] OR [rtol=56.342, atol=1e-05] 

to

      [rtol=1e-05, atol=1.9761e-05] OR [rtol=1.251e-05, atol=1e-05]

So yes some fused kernel has accuracy issue in your network. 

I notice you are run TRT 7 with cublas 10.1, could you update you cublas to 10.2.3.254 to solve the warning

      [TensorRT] WARNING: TensorRT was linked against cuBLAS 10.2.2 but loaded cuBLAS 10.1.0

10.1.0 is not in the support matrix, see https://docs.nvidia.com/deeplearning/tensorrt/support-matrix/index.html, the download link is in https://developer.nvidia.com/cuda-10.2-download-archive?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1804&target_type=runfilelocal

**Comment by taxuezcy** [comment]
hello, i use cuda10.2  not cuda10.1

**Comment by ttyio** [comment]
@taxuezcy , you cuda version is 10.2 but seems your cublas is still 10.1 which is not in TRT support matrix.
could you enable verbose log and the build the engine again, if cublas tactic is selected, it is probably cublas compatibility issue. We cannot use cublas10.1 because QA test only cover the configuration listed in the support matrix.

**Comment by taxuezcy** [comment]
@ttyio ,this is my cublas library 

/usr/lib/x86_64-linux-gnu/libcublas.so
/usr/lib/x86_64-linux-gnu/libcublas.so.10
/usr/lib/x86_64-linux-gnu/libcublas.so.10.2.2.89
i don't know why this warning happened, how to solve it ,thanks

**Comment by ttyio** [comment]
@taxuezcy , could you paste the output of 

        ls -al /usr/lib/x86_64-linux-gnu/libcublas.so.10

here, also from what I know, old 10.2 version has known issue, that's why we use Patch1(10.2.2.214) in TRT 7.2.1 and Patch2 (10.2.3.254) in TRT 7.2.2 in the support matrix.

What about first install the patch in https://developer.nvidia.com/cuda-10.2-download-archive?target_os=Linux&target_arch=x86_64&target_distro=Ubuntu&target_version=1804&target_type=runfilelocal and let's retry first?
thanks!

**Comment by taxuezcy** [comment]
@ttyio , the output of

    ls -al /usr/lib/x86_64-linux-gnu/libcublas.so.10 is /usr/lib/x86_64-linux-gnu/libcublas.so.10 -> libcublas.so.10.2.2.89

**Comment by ttyio** [comment]
@taxuezcy 
Hmmm, the version is reported from API `cublasGetVersion`, have you set `LD_LIBRARY_PATH` or `LD_PRELOAD` in your system? another way is to break in gdb and `info shared` to check which cublas library is loaded.

**Comment by taxuezcy** [comment]
@ttyio ,you see 
[W] 'colored' module is not installed, will not use colors when logging. To enable colors, please install the 'colored' module: python3 -m pip install colored
[I] Runner: trt-runner-N0-02/26/21-11:22:06          | Activating and starting inference
[TensorRT] ERROR: Parameter check failed at: ../builder/BuilderConfig.cpp::clearFlag::44, condition: int(builderFlag) >= 0 && int(builderFlag) < EnumMax<BuilderFlag>()
[TensorRT] ERROR: Parameter check failed at: ../builder/BuilderConfig.cpp::getFlag::27, condition: int(builderFlag) >= 0 && int(builderFlag) < EnumMax<BuilderFlag>()
[I] Building engine with configuration: max_workspace_size=16777216 bytes (16.00 MB) | tf32=False, fp16=False, int8=False, strict_types=False | 1 profiles
[TensorRT] WARNING: Current optimization profile is: 0. Please ensure there are no enqueued operations pending in this context prior to switching profiles
[I] Runner: trt-runner-N0-02/26/21-11:22:06          | Completed 1 iterations.
[I] Runner: onnxrt-runner-N0-02/26/21-11:22:06       | Activating and starting inference
[I] Runner: onnxrt-runner-N0-02/26/21-11:22:06       | Completed 1 iterations.
[I] Accuracy Comparison | trt-runner-N0-02/26/21-11:22:06 vs. onnxrt-runner-N0-02/26/21-11:22:06
[I]     Comparing Output: '534' (dtype=float32, shape=(1, 1, 128, 128)) with '534' (dtype=float32, shape=(1, 1, 128, 128))
[I]         Required tolerances: [atol=1.7366] OR [rtol=1e-05, atol=1.7365] OR [rtol=0.44963, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-11:22:06          | Stats: mean=-5.3844, min=-7.9094 at (0, 0, 115, 3), max=-2.4424 at (0, 0, 126, 126)
            Runner: onnxrt-runner-N0-02/26/21-11:22:06       | Stats: mean=-5.3046, min=-8.0653 at (0, 0, 49, 9), max=-2.7 at (0, 0, 77, 0)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[I]     Comparing Output: '535' (dtype=float32, shape=(1, 2, 128, 128)) with '535' (dtype=float32, shape=(1, 2, 128, 128))
[I]         Required tolerances: [atol=0.23577] OR [rtol=1e-05, atol=0.23577] OR [rtol=5.8152, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-11:22:06          | Stats: mean=0.43141, min=-0.046996 at (0, 0, 50, 2), max=0.9565 at (0, 0, 35, 126)
            Runner: onnxrt-runner-N0-02/26/21-11:22:06       | Stats: mean=0.41433, min=-0.018458 at (0, 0, 50, 2), max=0.89764 at (0, 1, 121, 96)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[I]     Comparing Output: '536' (dtype=float32, shape=(1, 2, 128, 128)) with '536' (dtype=float32, shape=(1, 2, 128, 128))
[I]         Required tolerances: [atol=4.1119] OR [rtol=1e-05, atol=4.1117] OR [rtol=56.342, atol=1e-05]
            Runner: trt-runner-N0-02/26/21-11:22:06          | Stats: mean=2.289, min=0.41181 at (0, 1, 127, 114), max=20.953 at (0, 1, 66, 1)
            Runner: onnxrt-runner-N0-02/26/21-11:22:06       | Stats: mean=2.2614, min=-0.12555 at (0, 0, 94, 0), max=23.085 at (0, 1, 65, 1)
[E]         FAILED | Difference exceeds tolerance (rtol=1e-05, atol=1e-05)
[E]     FAILED | Mismatched outputs: ['534', '535', '536']
after set LD_LIBRARY_PATH,without the cublas warning,but the result is still wrong

**Comment by ttyio** [comment]
Hello @taxuezcy , could you update to TRT 7.2.2 and retry? We have many fixes in each version. 
If you still observe mismatch in 7.2.2, run again with:

        --trt-outputs mark all --onnx-outputs mark all 

check the ONNX model to see which layer produced the first mismatched tensor.

**Comment by taxuezcy** [comment]
@ttyio ,i change to trt7.2.2 ,and the accuracy loss does better than trt7.0 ,but still loss compared with origin pytorch model

**Comment by ttyio** [comment]
@taxuezcy , the mismatch doesn't have to be accuracy loss, we cannot avoid arithmetic error for floating point. And polygraphy is a debug tool using synesthetic data, if you feed the model with the real input, how is the end to end accuracy?

**Comment by taxuezcy** [comment]
hi,   what i say is the end to end accuracy has a little loss when compared with origin pytorch model, @ttyio

**Comment by ttyio** [comment]
@taxuezcy , for accuracy drop, which metric are you using and what is the number?

We can debug the accuracy either using polygraphy or manually, 

- to debug using polygraphy, use  `--trt-outputs mark all --onnx-outputs mark all ` and find out the earliest mismatched tensor, then check the onnx model to find out the produce layer. The cons for this approach is that mark all layers as output would break tensorRT's layer fusion, so we may miss the bugs caused by fused kernel.

- to manually debug, we can slice the onnx model manually, e.g, only leave [0, N/2] nodes from onnx model, mark last node as output node, run the model. We will slice and run many iterations until find the node that cause mismatch.

**Comment by ttyio** [comment]
Closing since no activity in this thread for more than 3 weeks, please reopen if you still have question, thanks!
